twilio_integration/twilio_integration/doctype/whatsapp_message/whatsapp_message.py

```diff
@@ -165,3 +184,131 @@ def are_whatsapp_messages_muted():
 
 def is_whatsapp_enabled():
 	return True if frappe.get_cached_value("Twilio Settings", None, 'enabled') else False
+
+def flush_wa_queue(queue=False):
+	"""Flush queued WhatsApp Messages, called from scheduler"""
+	auto_commit = not queue
+
+	if are_whatsapp_messages_muted():
+		frappe.msgprint(_("WhatsApp messages are muted"))
+		return
+
+	for message in get_queued_messages():
+		send_one_wa_message(message.name, auto_commit, queue=queue)
+
+def send_one_wa_message(message_name, auto_commit=True, queue=False):
+	if are_whatsapp_messages_muted():
+		frappe.msgprint(_("WhatsApp messages are muted"))
+		return
+
+	wa_message = frappe.db.sql('''
+		select *
+		from `tabWhatsApp Message`
+		where name = %s
+		for update
+	''', message_name, as_dict=True)[0]
+
+	if wa_message.status not in ('Not Sent', 'Queued'):
+		if auto_commit:
+			frappe.db.rollback()
+		return
+
+	frappe.db.sql("""
+		update `tabWhatsApp Message` set status='Sending', modified=%s where name=%s
+	""", (now_datetime(), wa_message.name), auto_commit=auto_commit)
+
+	if wa_message.communication:
+		frappe.get_doc('Communication', wa_message.communication).set_delivery_status(commit=auto_commit)
+
+	try:
+		wa_msg = frappe.get_doc('WhatsApp Message', wa_message.name)
+		wa_msg.send()
+
+		if wa_message.communication:
+			frappe.get_doc('Communication', wa_message.communication).set_delivery_status(commit=auto_commit)
+
+		run_after_send_method(wa_message.reference_doctype, wa_message.reference_document_name, wa_message.notification_type)
+
+	except (ConnectionError, Timeout, JobTimeoutException):
+		handle_timeout(wa_message, auto_commit)
+
+	except Exception as e:
+		handle_error(e, wa_message, auto_commit, queue)
+
+def get_queued_messages():
+	return frappe.db.sql('''
+		select name, from_
+		from `tabWhatsApp Message`
+		where status IN ('Queued', 'Not Sent')
+		order by priority desc, creation asc
+		limit 500
+	''', as_dict=True)
+
+def clear_wa_message_queue():
+	"""Remove WhatsApp messages older than 31 days with final status,
+	and expire WhatsApp messages not sent for 7 days.
+	Called daily via scheduler.
+	"""
+	final_statuses = ("Sent", "Delivered", "Read", "Failed", "Undelivered", "Not Sent")
+
+	old_whatsapp_messages = frappe.db.sql_list("""
+		SELECT name
+		FROM `tabWhatsApp Message`
+		WHERE status IN %s
+		AND modified < (NOW() - INTERVAL '31' DAY)
+	""", [final_statuses])
+
+	if old_whatsapp_messages:
+		frappe.db.sql("""
+			DELETE FROM `tabWhatsApp Message`
+			WHERE name IN %s
+		""", [old_whatsapp_messages])
```

Comment:
do not delete WhatsApp Message, WhatsApp Message is not supposed to be temporary